packages/core/core/src/Graph.js

```diff
@@ -85,6 +85,20 @@ export default class Graph<TNode: Node> {
     return this.outboundEdges.get(from).has(to);
   }
 
+  ensureConnection(fromNode: TNode, toNode: TNode) {
```

Comment:
I don't know how I feel about this method in general. It feels like using it is a sign that we're not sure the graph is in a good state.

Also, re: the name, `ensureConnection` feels like it should just make sure that the two nodes are connected in the graph (that behavior is pretty much what `addEdge` does now, as currently it's okay to re-add an edge). Something with this name actually adding missing nodes to the graph feels unexpected.

Can this behavior be moved into the call sites that are actually uncertain about the graph state? I feel like keeping this around might encourage bad behavior.

Reply:
>  It feels like using it is a sign that we're not sure the graph is in a good state.
I'm using it for the case "These two nodes may already be connected, but if they're not, connect them". I wouldn't call that a bad state. `replaceNodesConnectedTo` does this as well, but it also does some stuff I don't want like overwriting the value of the node that already existed and replacing all of the other nodes unless I provide a function. I guess `addEdge` ensures against duplicate edges now (it didn't before, when I originally wrote this code). I could remove the abstraction and change the call sites back to:

```
if (!this.hasNode(toNode)) {
  this.addNode(toNode)
}

this.addEdge(fromNodeId, toNodeId)
```

but I don't totally understand what makes the abstraction dangerous.